src/realm/data_type.hpp

```diff
@@ -151,7 +151,12 @@ static constexpr DataType type_OldTable = DataType{5};
 static constexpr DataType type_OldDateTime = DataType{7};
 static_assert(!type_OldTable.is_valid());
 static_assert(!type_OldDateTime.is_valid());
+
+// Non primitive types
 static constexpr DataType type_TypeOfValue = DataType{18};
+static constexpr DataType type_List = DataType{19};
+static constexpr DataType type_Set = DataType{20};
+static constexpr DataType type_Dictionary = DataType{21};
```

Comment:
Don't you have to change also the union for `struct DataType {` and methods like, 
`constexpr bool is_valid() const noexcept` ... or do you want to keep this types "hidden" on purpose?